signac/synced_collections/utils.py

```diff
@@ -111,18 +103,26 @@ def default(o: Any) -> Dict[str, Any]:  # noqa: D102
       serialized data may be incorrect.
 
     """
-    if NUMPY:
-        if isinstance(o, numpy.ndarray):
-            if o.shape == ():
-                return o.item()
-            else:
-                return o.tolist()
-        elif isinstance(o, numpy.number):
-            return o.item()
-    try:
-        return o._data
-    except AttributeError as e:
-        raise TypeError from e
+    # Numpy converters return the data unchanged.
+    converted_o = _convert_numpy(o)
+
+    # Numpy arrays will be converted to lists, then recursively parsed by the
+    # JSON encoder, so we only have to handle the case where we have a scalar
+    # type at the bottom level that can't been converted to a Python scalar.
+    if _is_numpy_scalar(converted_o):
+        raise ValueError(
+            "Only NumPy types that can be converted to raw Python types can be"
+            "JSON-encoded. All other types, such as NumPy extended-precision types, "
+            "must be converted by the user."
```

Comment:
```suggestion
            "Only NumPy types with corresponding Python types can be JSON-encoded. "
            "All other types, such as NumPy extended precision types, must be "
            "converted by the user."
```

Reply:
Exception to note: Complex values _do_ have corresponding Python types but are not JSON-encodable. See comment below.

Reply:
I applied your changes. I'll leave this comment unresolved for the moment and we can discuss on the other thread.

Reply:
Please add some clarifying description because complex dtypes do have corresponding types in Python but are not JSON-encodable, then this can be resolved.

Reply:
I rewrote this error to explicitly state that the existence of a conversion is a necessary but not sufficient condition.